yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr

```diff
@@ -0,0 +1,419 @@
+        storage.public_balances.at(to.address).write(to_balance.value as Field);
+
+        1
+    }
+
+    #[aztec(public)]
+    fn burn_public(
+        from: AztecAddress,
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Context::public(&mut context));
+
+        if (from.address != context.msg_sender()) {
+            let selector = compute_selector("burn_public((Field),Field,Field)");
+            let message_field = compute_message_hash([context.msg_sender(), context.this_address(), selector, from.address, amount, nonce]);
+            AccountContract::at(from.address).is_valid(Context::public(&mut context), message_field);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let amount = SafeU120::new(amount);
+        let from_balance = SafeU120::new(storage.public_balances.at(from.address).read()).sub(amount);
+        storage.public_balances.at(from.address).write(from_balance.value as Field);
+
+        let new_supply = SafeU120::new(storage.total_supply.read()).sub(amount);
+        storage.total_supply.write(new_supply.value as Field);
+
+        1
+    }
+
+    #[aztec(private)]
+    fn redeem_shield(
+        to: AztecAddress,
+        amount: Field,
+        secret: Field,
+    ) -> Field {
+        let storage = Storage::init(Context::private(&mut context));
+        let pending_shields = storage.pending_shields;
+        let balance = storage.balances.at(to.address);
+        let public_note = TransparentNote::new_from_secret(amount, secret);
+        
+        pending_shields.assert_contains_and_remove_publicly_created(public_note);
+        increment(balance, amount, to.address);
+
+        1
+    }
+
+    #[aztec(private)]
+    fn unshield(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Context::private(&mut context));
+
+        if (from.address != context.msg_sender()) {
+            let selector = compute_selector("unshield((Field),(Field),Field,Field)");
+            let message_field = compute_message_hash([context.msg_sender(), context.this_address(), selector, from.address, to.address, amount, nonce]);
+            AccountContract::at(from.address).is_valid(Context::private(&mut context), message_field);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let from_balance = storage.balances.at(from.address);
+        decrement(from_balance, amount, from.address);
+
+        let selector = compute_selector("_increase_public_balance((Field),Field)");
+        let _void = context.call_public_function(context.this_address(), selector, [to.address, amount]);
+
+        1
+    }
+
+    #[aztec(private)]
+    fn transfer(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
```

Comment:
Should the `amount`s in all functions eventually be a safer type than `Field`, i.e. some kind of uint (once Noir is changed to prevent under/overflows?)